src/zcash/Note.hpp

```diff
@@ -40,24 +42,55 @@ class SproutNote : public BaseNote {
     uint256 nullifier(const SproutSpendingKey& a_sk) const;
 };
 
+inline bool plaintext_version_is_valid(const Consensus::Params& params, int height, unsigned char leadbyte) {
+    int canopyActivationHeight = params.vUpgrades[Consensus::UPGRADE_CANOPY].nActivationHeight;
 
+    if (height < canopyActivationHeight && leadbyte != 0x01) {
+        // non-0x01 received before Canopy activation height
+        return false;
+    }
+    if (height >= canopyActivationHeight
+        && height < canopyActivationHeight + ZIP212_GRACE_PERIOD
+        && leadbyte != 0x01
+        && leadbyte != 0x02)
+    {
+        // non-{0x01,0x02} received after Canopy activation and before grace period has elapsed
+        return false;
+    }
+    if (height >= canopyActivationHeight + ZIP212_GRACE_PERIOD && leadbyte != 0x02) {
+        // non-0x02 received past (Canopy activation height + grace period)
+        return false;
+    }
+    return true;
```

Comment:
```suggestion
    switch (leadbyte) {
    case 0x01:
        return height < canopyActivationHeight + ZIP212_GRACE_PERIOD;
    case 0x02:
        return height >= canopyActivationHeight;
    default:
        return false;
    }
```

Reply:
this looks good to me; just a note that the current implementation is written to mirror the spec as closely as possible. maybe @daira would want to weigh in on this?

![](https://i.imgur.com/sAOnidW.png)